firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,479 @@
+                        .getInterface(Ci.nsIDOMWindowUtils)
+                        .outerWindowID.toString();
+            """)
+
+    def close(self, handle):
+        """Closes the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+
+        :returns: TODO
+        """
+        self.switch_to(handle)
+
+        # TODO: Maybe needs to wait as handled via an observer
+        return self.marionette.close_chrome_window()
+
+    def close_all(self, exceptions=None):
+        """Closes all open chrome windows.
+
+        There is an optional `exceptions` list, which can be used to exclude
+        specific chrome windows from being closed.
+
+        :param exceptions: Optional, list of handles or :class:`BaseWindow` instances
+        not to close
+        """
+        handles_to_keep = exceptions or []
+        handles_to_keep = [entry.handle if isinstance(entry, BaseWindow) else entry
+                           for entry in exceptions]
+
+        for handle in self.marionette.chrome_window_handles:
+            if handle not in handles_to_keep:
+                self.close(handle)
+
+    def create_window_instance(self, handle, expected_class=None):
+        """Creates a :class:`BaseWindow` instance for the given chrome window
+
+        :param handle: The handle of the chrome window
+        :param expected_class: Optional, check for the correct window class
+        """
+        current_handle = self.marionette.current_chrome_window_handle
+
+        try:
+            # Retrieve window type to determine the type of window
+            self.switch_to(handle)
+            window_type = self.marionette.get_window_type()
+        finally:
+            # Ensure to switch back to the original window
+            if handle != current_handle:
+                self.marionette.switch_to_window(current_handle)
+
+        window = BaseWindow(lambda: self.marionette, handle)
+
+        if window_type == 'navigator:browser':
+            window.__class__ = BrowserWindow
+        else:
+            raise errors.UnknownWindowError('Unknown window type "%s" for handle: "%s"' %
+                                            (window_type, handle))
+
+        if expected_class is not None and type(window) is not expected_class:
+            raise errors.UnexpectedWindowTypeError('Expected window "%s" but got "%s"' %
+                                                   (expected_class, type(window)))
+
+        return window
+
+    def focus(self, handle):
+        """Focuses the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+        """
+        self.switch_to(handle)
+
+        with self.marionette.using_context('chrome'):
+            self.marionette.execute_script(""" window.focus(); """)
+
+        wait = Wait(self.marionette, timeout=5, interval=.1)
+        wait.until(lambda m: handle == self.focused_chrome_window_handle)
```

Comment:
switch_to() does not set the focus to the newly selected chrome window or tab. The focus will still be on the old handle. Otherwise not sure what you are asking here.